File: python/src/core/polyline.py
from core.OpenFigure import OpenFigure
# from point import Point
# from point2d import Point2D
# from segment import Segment

class Polyline(OpenFigure):
	def __init__(self, n, p):
		self.n = n
		self.p = p

	def __str__(self):
		st = f"Polyline: size = {self.n}; "
		for i in range(len(self.p)):
			st += f"{str(self.p[i])}, "
		st = st[:len(st)-2]
		return st

	def getP(self, i = None):
		if i == None:
			return self.p
		else:
			return self.p[i]

	def setP(self, newP, i = None):
		if i == None:
			self.p = newP
		else:
			self.p[i] = newP

	def length(self):
		l = 0
		pprev = self.p[0]
		for point in self.p:
			l += Segment(pprev, point).length()
			pprev = point
		return l

	def shift(self, a):
		result = polyline(self.n, self.p)
		for ns in range(result.n):
			result.p[ns] = result.p[ns].add(a)
		return result

	def rot(self, phi):
		result = polyline(self.n, self.p)
		for ns in range(result.n):
			result.p[ns] = result.p[ns].rot(phi)
		return result

	def symAxis(self, i):
		if i < 0:
			raise Exception("Ошибка: отрицательное пространство")
		if self.p[0].dim < i:
			raise Exception("Ошибка: имеет пространство меньшее, чем нужное")
		result = Polyline(self.n, self.p)
		for ns in range(result.n):
			result.p[ns] = result.p[ns].symAxis(i)
		return result

	def cross(self, figure):
		pprev = self.p[0]
		for pt in self.p[1:]:
			if Segment(pprev, pt).cross(figure):
				return True
			pprev = pt
		return False

File: csharp/WorkOOP/Polyline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WorkOOP
{
    // Реализация класса Polyline – ломаная линия
    internal class Polyline: OpenFigure
    {
        protected int n;        // n – количество точек.
        protected Point2D[] p;  // p – массив двумерных точек.

        /*
         * Polyline – конструктор, описывающий инициализацию объекта.
         * 
         * На вход принимает массив точек в порядке обхода по ломаной.
         */
        public Polyline(Point2D[] p) {
            if (p.Length == 0)
                throw new Exception("No zero array polyline!!");
            n = p.Length;
            this.p = p; 
        }

        /*
         * Реализация всех функций из OpenFigure, кроме square() 
         */
        public int getN() => n;
        public Point2D[] getP() => p;
        public Point2D getP(int i) => p[i];
        public void setP(Point2D[] p) { this.p = p; n = p.Length; }
        public void setP(Point2D p, int i) => this.p[i] = p;

        public override double length() {
            double l = 0;
            Point2D prevp = p[0];
            foreach (Point2D point in p) { l += new Segment(prevp, point).length(); prevp = point; }
            return l;
        }

        public override Polyline shift(Point2D a) {
            foreach (Point2D e in p)
                e.add(a);
            return this; 
        }

        public override Polyline rot(double phi) {
            foreach (Point2D e in p)
                e.rot(phi);
            return this; 
        }

        public override Polyline symAxis(int i) {
            foreach (Point2D e in p)
                e.symAxis(i);
            return this; 
        }

        public override bool cross(IShape i) {
            Point2D prev = p[0];
            foreach (Point2D pt in p)
            {
                if (new Segment(prev, pt).cross(i))
                    return true;
                prev = pt;
            }
            return false;
        }

        public override String toString() { 
            StringBuilder sb = new StringBuilder(); sb.Append("Size: " + n + "\nPoints: [ ");
            foreach (Point2D e in p)
                sb.Append(e.toString() + " ");
            sb.Append(" ]");
            return sb.ToString();
        }
    }
}
